File: packages/contracts/src/launchdarkly.ts
import { z } from 'zod';

export const FlagVariationSchema = z.object({
	id: z.string(),
	name: z.string().optional(),
	value: z.unknown(),
});

export const FlagRuleSchema = z.object({
	id: z.string(),
	description: z.string().optional(),
	clauses: z.array(z.unknown()),
	variationOrRollout: z.unknown(),
});

export const FlagEnvironmentStateSchema = z.object({
	on: z.boolean(),
	rules: z.array(FlagRuleSchema),
	fallthrough: z.unknown(),
	offVariation: z.number().optional(),
	targets: z.array(z.unknown()).optional(),
});

export const FlagStateSchema = z.object({
	projectKey: z.string(),
	flagKey: z.string(),
	variations: z.array(FlagVariationSchema),
	environments: z.record(z.string(), FlagEnvironmentStateSchema),
});

export const VariationRefSchema = z.discriminatedUnion('by', [
	z.object({ by: z.literal('id'), id: z.string() }),
	z.object({ by: z.literal('name'), name: z.string() }),
	z.object({ by: z.literal('value'), value: z.unknown() }),
]);

export const RuleRefSchema = z.discriminatedUnion('by', [
	z.object({ by: z.literal('id'), id: z.string() }),
	z.object({ by: z.literal('description'), description: z.string() }),
]);

export const RolloutIntentSchema = z.object({
	mode: z.enum(['fallthrough', 'rule']).default('fallthrough'),
	treatmentVariationRef: VariationRefSchema,
	controlVariationRef: VariationRefSchema,
	treatmentPercentThousandths: z.number().int().min(0).max(100_000),
	rolloutContextKind: z.string(),
	rolloutBucketBy: z.string(),
	ruleRef: RuleRefSchema.optional(),
});

export const SemanticPatchInstructionSchema = z.discriminatedUnion('kind', [
	z.object({ kind: z.literal('turnFlagOn') }),
	z.object({ kind: z.literal('turnFlagOff') }),
	z.object({
		kind: z.literal('updateFallthroughVariationOrRollout'),
		rolloutWeights: z.record(z.string(), z.number().int()),
		rolloutBucketBy: z.string(),
		rolloutContextKind: z.string(),
	}),
	z.object({
		kind: z.literal('updateRuleVariationOrRollout'),
		ruleId: z.string(),
		rolloutWeights: z.record(z.string(), z.number().int()),
		rolloutBucketBy: z.string(),
		rolloutContextKind: z.string(),
	}),
]);

export const TargetingIntentSchema = z.object({
	environmentKey: z.string(),
	comment: z.string().optional(),
	turnOn: z.boolean().optional(),
	rollout: RolloutIntentSchema.optional(),
});

export const GetFlagStateInputSchema = z.object({
	projectKey: z.string(),
	flagKey: z.string(),
	environmentKey: z.string(),
});

export const ApplyTargetingInputSchema = z.object({
	projectKey: z.string(),
	flagKey: z.string(),
	intent: TargetingIntentSchema,
});

export const LaunchDarklyClientConfigSchema = z.object({
	accessToken: z.string(),
	baseUrl: z.string().url().optional(),
	apiVersion: z.string().optional(),
});

export type FlagVariation = z.infer<typeof FlagVariationSchema>;
export type FlagRule = z.infer<typeof FlagRuleSchema>;
export type FlagEnvironmentState = z.infer<typeof FlagEnvironmentStateSchema>;
export type FlagState = z.infer<typeof FlagStateSchema>;
export type VariationRef = z.infer<typeof VariationRefSchema>;
export type RuleRef = z.infer<typeof RuleRefSchema>;
export type RolloutIntent = z.infer<typeof RolloutIntentSchema>;
export type SemanticPatchInstruction = z.infer<
	typeof SemanticPatchInstructionSchema
>;
export type TargetingIntent = z.infer<typeof TargetingIntentSchema>;
export type GetFlagStateInput = z.infer<typeof GetFlagStateInputSchema>;
export type ApplyTargetingInput = z.infer<typeof ApplyTargetingInputSchema>;
export type LaunchDarklyClientConfig = z.infer<
	typeof LaunchDarklyClientConfigSchema
>;

File: kotlin/modules/contracts/bin/main/com/ffpromo/contracts/LaunchDarkly.kt
package com.ffpromo.contracts

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import kotlinx.serialization.json.JsonElement

@Serializable
data class FlagVariation(
    val id: String,
    val name: String? = null,
    val value: JsonElement,
)

@Serializable
data class FlagRule(
    val id: String,
    val description: String? = null,
    val clauses: List<JsonElement> = emptyList(),
    val variationOrRollout: JsonElement? = null,
)

@Serializable
data class FlagEnvironmentState(
    val on: Boolean,
    val rules: List<FlagRule> = emptyList(),
    val fallthrough: JsonElement? = null,
    val offVariation: Int? = null,
    val targets: List<JsonElement>? = null,
)

@Serializable
data class FlagState(
    val projectKey: String,
    val flagKey: String,
    val variations: List<FlagVariation>,
    val environments: Map<String, FlagEnvironmentState>,
)

@Serializable
@JsonClassDiscriminator("by")
sealed class VariationRef {
    @Serializable
    @SerialName("id")
    data class ById(val id: String) : VariationRef()

    @Serializable
    @SerialName("name")
    data class ByName(val name: String) : VariationRef()

    @Serializable
    @SerialName("value")
    data class ByValue(val value: JsonElement) : VariationRef()
}

@Serializable
@JsonClassDiscriminator("by")
sealed class RuleRef {
    @Serializable
    @SerialName("id")
    data class ById(val id: String) : RuleRef()

    @Serializable
    @SerialName("description")
    data class ByDescription(val description: String) : RuleRef()
}

@Serializable
data class RolloutIntent(
    val mode: String = "fallthrough",
    val treatmentVariationRef: VariationRef,
    val controlVariationRef: VariationRef,
    val treatmentPercentThousandths: Int,
    val rolloutContextKind: String,
    val rolloutBucketBy: String,
    val ruleRef: RuleRef? = null,
)

@Serializable
@JsonClassDiscriminator("kind")
sealed class SemanticPatchInstruction {
    @Serializable
    @SerialName("turnFlagOn")
    data object TurnFlagOn : SemanticPatchInstruction()

    @Serializable
    @SerialName("turnFlagOff")
    data object TurnFlagOff : SemanticPatchInstruction()

    @Serializable
    @SerialName("updateFallthroughVariationOrRollout")
    data class UpdateFallthroughVariationOrRollout(
        val rolloutWeights: Map<String, Int>,
        val rolloutBucketBy: String,
        val rolloutContextKind: String,
    ) : SemanticPatchInstruction()

    @Serializable
    @SerialName("updateRuleVariationOrRollout")
    data class UpdateRuleVariationOrRollout(
        val ruleId: String,
        val rolloutWeights: Map<String, Int>,
        val rolloutBucketBy: String,
        val rolloutContextKind: String,
    ) : SemanticPatchInstruction()
}

@Serializable
data class TargetingIntent(
    val environmentKey: String,
    val comment: String? = null,
    val turnOn: Boolean? = null,
    val rollout: RolloutIntent? = null,
)

@Serializable
data class GetFlagStateInput(
    val projectKey: String,
    val flagKey: String,
    val environmentKey: String,
)

@Serializable
data class ApplyTargetingInput(
    val projectKey: String,
    val flagKey: String,
    val intent: TargetingIntent,
)

@Serializable
data class LaunchDarklyClientConfig(
    val accessToken: String,
    val baseUrl: String? = null,
    val apiVersion: String? = null,
)
